WordPressAuthenticator/OAuth/ProofKeyForCodeExchange.swift

```diff
@@ -25,21 +25,92 @@ struct ProofKeyForCodeExchange {
         }
     }
 
-    let codeVerifier: String
+    let codeVerifier: CodeVerifier
     let method: Method
 
-    init(codeVerifier: String, method: Method) {
+    init() throws {
+        self.codeVerifier = try .makeRandomCodeVerifier()
+        self.method = .s256
+    }
+
+    init(codeVerifier: CodeVerifier, method: Method) {
         self.codeVerifier = codeVerifier
         self.method = method
     }
 
     var codeCallenge: String {
         switch method {
         case .s256:
-            // TODO: code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
-            fatalError()
+            // The spec defines code_challenge for the s256 mode as:
+            //
+            // code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
+            //
+            // We don't need the ASCII conversion, because we build `CodeVerifier` from URL safe
+            // characters.
+            let rawData = Data(codeVerifier.rawValue.utf8)
+            let hashedData: Data = rawData.sha256Hashed()
+            return hashedData.base64URLEncodedString()
         case .plain:
-            return codeVerifier
+            return codeVerifier.rawValue
+        }
```

Comment:
Currently the `CodeVerifier` type doesn't too much, I guess it's like a glorified string (with some additional checks). However, I think we can make use of it by moving this code challenge calculation (BTW, there is a typo in the variable name: C"h"allenge) into the CodeVerifier, for example `CodeVerifier.codeChallenge(using method: Method)`. Then, the hashing and base 64 encoding can be encapsulated inside the `CodeVerifier` type, along with some specification around the code verifier. What do you think?

Reply:
Great idea!

Reply:
02cfa0a2